Source/CloudyPanelPlugin/Private/CloudyPanelPlugin.cpp

```diff
@@ -90,6 +122,59 @@ bool CCloudyPanelPluginModule::Tick(float DeltaTime)
 	return true;
 }
 
+
+bool CCloudyPanelPluginModule::CaptureFrame(float DeltaTime) {
+	//UE_LOG(ModuleLog, Warning, TEXT("Testing capture frame"));
+	UE_LOG(ModuleLog, Warning, TEXT("time %f"), DeltaTime); // can track running time
+	
+	if (!isEngineRunning && GEngine->GameViewport != nullptr && GIsRunning && IsInGameThread()) { // engine has been started
+		isEngineRunning = true;
+		CCloudyPanelPluginModule::SetUpVideoCapture();
+		UE_LOG(ModuleLog, Warning, TEXT("engine started"));
+	}
+	else if (isEngineRunning && !(GEngine->GameViewport != nullptr && GIsRunning && IsInGameThread())) { // engine has been stopped
+		isEngineRunning = false;
+		UE_LOG(ModuleLog, Warning, TEXT("engine stopped"));
+		
+		fclose(VideoPipe);
+	}
+
+	if (GEngine->GameViewport != nullptr && GIsRunning && IsInGameThread())
+	{
+		
+		// init buffers for each player, up to 4
+		TArray<FColor> OutBuffer, OutBuffer1, OutBuffer2, OutBuffer3, OutBuffer4;
+		
+		// get viewport and split screen data
+		FViewport* ReadingViewport = GEngine->GameViewport->Viewport;
+		ReadingViewport->ReadPixels(FrameBuffer);
+		ULocalPlayer* FirstPlayer = GEngine->FindFirstLocalPlayerFromControllerId(0);
+		TArray<FSplitscreenData> SplitscreenInfo = FirstPlayer->ViewportClient->SplitscreenInfo;
+		int SplitscreenType = FirstPlayer->ViewportClient->GetCurrentSplitscreenConfiguration();
+
+		// use VideoPipe (class variable) to pass frames to encoder
+		int sizeX = ReadingViewport->GetSizeXY().X;
+		int sizeY = ReadingViewport->GetSizeXY().Y;
+		uint32 *PixelBuffer;
+		PixelBuffer = new uint32[sizeX * sizeY * PIXEL_SIZE];
+		
+		int i = 0;
+		for (auto& Pixel : FrameBuffer) {
+			std::ostringstream PixelStream;
+			PixelBuffer[i] = Pixel.A * 256 * 256 * 256 + Pixel.B * 256 * 256 + Pixel.G * 256 + Pixel.R;
```

Comment:
I'm not sure how much faster bit shift is, but you can try:
```
PixelBuffer[i] = Pixel.A << 24 | Pixel.B << 16 | Pixel.G << 8 | Pixel.R;
```
It doesn't seem to have any difference when I tried. Maybe just keep this idea in mind.